### ISSUE TITLE
Audio and hotkeys don't work sporadically

### ISSUE THREAD

**Issue by GDur** [mumble-voip/mumble#4992]
Sometimes another person in mumble can talk to me, but my push to talk hotkey doesn't work anymore, so I can't reply.
And sometimes mumble doesn't relay the audio of other channel members on top of that. (A restart fixes the problem)

(In these cases, the "voice-indicator-lamp" is also not reacting)

 - OS:  Windows 10
 - Version: Mumble 1.3.4 (But the problem exists for a long time)

What exactly triggers this behavior is not clear to me, but if I had a guess then that it happens after some time of being inactive.
However a friend of mine said he experienced it directly after starting mumble.

Thanks for reading :)

**Comment by vimpostor** [comment]
> Thanks for reading :)

Thanks for reporting.

Perhaps this is again a case of someone not using the Opus codec and throwing the whole session off.

> However a friend of mine said he experienced it directly after starting mumble.

Does this mean your friend experiences the problem or you have the problem immediately after your friend joins? In any case, is that friend also using the latest version of Mumble?

**Comment by Krzmbrzl** [comment]
Could you switch to Voice-activation or maybe even continous transmission and check whether the issue can still occur with that?
Furthermore could you set up another shortcut and check whether that is still working when PTT is not? Use some action that is independent of any audio functionality working as expected in order to really test whether this is shortcut- or audio-related.

**Comment by GDur** [comment]
> Does this mean your friend experiences the problem or you have the problem immediately after your friend joins? 

Yes.
> In any case, is that friend also using the latest version of Mumble?

Yes.

> Could you switch to Voice-activation or maybe even continious transmission and check whether the issue can still occur with that?

This is a bit too cumbersome for me to be honest as I'm using mumble quite frequently and used to using PTT.

> Furthermore could you set up another shortcut and check whether that is still working when PTT is not? Use some action that is independent of any audio functionality working as expected in order to really test whether this is shortcut- or audio-related.

I already have multiple PTT-hotkeys and I'm pretty sure neither of them is working when this issue occurs. As the "send-audio-indicator-lamp" is not "lighting up" i'm pretty confident, that the hotkeys don't work.

Hints:
Yet another friend told me that simply opening the mumble settings and pressing the OK-button (instead of restarting mumble) fixes the problem too. I tried it just now and it worked. 
He added then that he thinks this could be linked to the standby mode. As in: maybe mumbles hotkeys don't work properly after the PC/Laptop was in standby mode.

**Comment by Krzmbrzl** [comment]
> > Does this mean your friend experiences the problem or you have the problem immediately after your friend joins?

> Yes.

You can't answer an "either, or" question with "yes" xD

> This is a bit too cumbersome for me to be honest as I'm using mumble quite frequently and used to using PTT.

Okay then, but without more insights into the issue there is no way we can fix this :shrug: 

> I already have multiple PTT-hotkeys and I'm pretty sure neither of them is working when this issue occurs. As the "send-audio-indicator-lamp" is not "lighting up" i'm pretty confident, that the hotkeys don't work.

Unfortunately that does not help. That's why I wrote
> Use some action that is independent of any audio functionality working as expected in order to really test whether this is shortcut- or audio-related.

With this information it could still be that the audio system got hung up. In that case the PTT shortcut could be working just fine but if the underlaying audio system does not collaborate, then there still is no audio that is being sent (and thus also no speech indicator lighting up).

**Comment by GDur** [comment]
> You can't answer an "either, or" question with "yes" xD

That's true!😆 my bad. Yes sometimes he experiences the problem too.

> Okay then, but without more insights into the issue there is no way we can fix this 🤷

That's what I was afraid of. The hints could help:

> Yet another friend told me that simply opening the mumble settings and pressing the OK-button (instead of restarting mumble) fixes the problem too

What happens when the OK-button of the settings menu is triggered which is able to fix the problem?

**Comment by Krzmbrzl** [comment]
> What happens when the OK-button of the settings menu is triggered which is able to fix the problem?

I know for sure that the audio system is restarted. I currently don't know whether we are also restarting the shortcut engine though :thinking:


I think I have found a way to check which system part is affected: Could you open the PTT window (AudioInput settings) and when your PTT button fails to work, check whether clicking the button in that window works instead.

**Comment by live4ever** [comment]
Do you RDP into your PC (resorted to using vnc-type remote access for now)? I've noticed audio stops working and always comes back after opening config/settings (audio system restart).
https://github.com/mumble-voip/mumble/issues/1085

**Comment by GDur** [comment]
Tiny update:
Since I activated "Display push to talk window" (ca. nine days ago) the issue did not occur again.

But I don't know if this is a coincident or not.

**Comment by Krzmbrzl** [comment]
> Since I activated "Display push to talk window" (ca. nine days ago) the issue did not occur again.

Are you know using that window actively or is it just open somewhere but you continue to use your regular PTT key? :thinking:

**Comment by GDur** [comment]
Hi Mumble Team,

now I have a more concrete hint and a reproducible error:

Starting Point: 
The PTT-Key works as expected.

1. Disables or disconnect the microphone which is active in Mumble.
2. The PTT-Key stops working
3. Enable or connect the microphone again
4. The PTT-Key will still not work
5. Restart Mumble (or open and close the Mumble settings) and the PTT-Key will work again.

Seems to me like Mumble needs a hook for the case that the mic is usable again and should enable the PTT-Key again.

For more questions I am available of course. Thanks for reading. 👍

**Comment by Krzmbrzl** [comment]
@GDur 
> Could you open the PTT window (AudioInput settings) and when your PTT button fails to work, check whether clicking the button in that window works instead.

My bet is that it's not the hotkey that stops working (I am not aware of a mechanism that would disable the PTT key once the mic is no longer available) but rather the audio system getting hung up as a whole.

**Comment by GDur** [comment]
> My bet is that it's not the hotkey that stops working (I am not aware of a mechanism that would disable the PTT key once the mic is no longer available) but rather the audio system getting hung up as a whole.

Might be a good distinction. Maybe it's not the PTT-Key.
The more precise description is: The voice is not being transmitted and the indicator "lamp" is off too.

**Comment by Krzmbrzl** [comment]
Yeah maybe - that's why I would like you to answer the question about whether the PTT window is able to correctly cause audio to be transmitted. If it does, it's a shortcut issue. If it doesn't, it's an audio issue.

**Comment by GDur** [comment]
> Yeah maybe - that's why I would like you to answer the question about whether the PTT window is able to correctly cause audio to be transmitted. If it does, it's a shortcut issue. If it doesn't, it's an audio issue.

Thank you for the clarification. More testing revealed that neither the PTT-Key nor the PTT window are able to enable sound transmission.


Additionally I've tested what happens if instead the **voice activation** is in use:
After removing/disabling the mic, the "Lamp" is always "on" but there is no sound transmission.
Even after enabling the mic, the "Lamp" is still always "on" and still no sound transmission possible.

**Comment by Krzmbrzl** [comment]
Alright, then this is not an issue with the shortcuts, but rather with the audio engine getting stuck (presumably during restarts)

**Comment by Mnior** [comment]
#4300
#6346

It seems to me that the problem is the same everywhere.
mumble does not respond adequately to changes in the list of input and output devices.
For example, when you connect via RDP, the list of Audio Out/In and keyboards changes, back and forth.

Moreover, if you go into the settings and save them, without even changing them, then everything is restored, because re-binding occurs.

**Comment by Mnior** [comment]
> changes in the list of input and output devices

If the system does not provide logical access to devices, even if they are physically present.
Then the question remains: What to do if there can be several environments?
RDP may work differently depending on the OS.

Ideally, it would be good if for each terminal it was possible to specify its own setting (for devices):
https://learn.microsoft.com/en-us/windows/win32/api/wtsapi32/nf-wtsapi32-wtsquerysessioninformationa
And if the setting is saved with RDP, then a new entry is created for it and tied to it.
And when connected, reconnect all devices according to the settings for the current console.
If there is no setting, use the default devices.

**Comment by Mnior** [comment]
> Ideally, it would be good if for each terminal it was possible to specify its own setting

Because this requires a lot of changes and rarely occurs, apparently it’s worth doing this: if a specific device is specified, at the moment of changes (change of equipment or console session or settings?) If this device is not there, use the default device (virtual).

**Comment by MatthewFagan** [comment]
Hi. I am suffering from what I believe is the same core issue. For me, the audio disappears in mumble whenever I run my DAW software (Cakewalk SONAR). When the DAW starts, it does some kind of reset to the audio devices which causes them to disconnect/reconnect. This causes momentary sound drop-outs in all other software (web browsers, audio players, etc.), but Mumble does not reconnect like other software does.

I looked through the source, and could not find any immediate problems. On running the developer console, I get the following log events when I start the DAW:
<W>2026-03-17 10:27:35.357 WASAPIOutput: GetCurrentPadding failed: hr=0x88890004
<W>2026-03-17 10:27:35.358 WASAPIInput: GetNextPacketSize failed: hr=0x88890004

I had thought it might be related to [IMMNotificationClient](https://learn.microsoft.com/en-us/windows/win32/api/Mmdeviceapi/nn-mmdeviceapi-immnotificationclient) events, but those do not seem to get triggered by starting the DAW.

I looked a bit further, and 0x88890004 corresponds to AUDCLNT_E_DEVICE_INVALIDATED. There is a whole page dedicated to explaining how to recover from this error: https://learn.microsoft.com/en-us/windows/win32/coreaudio/recovering-from-an-invalid-device-error. It seems that Mumble does not implement this correctly.

OK, so I can see the problem in the WASAPI.cpp file. On any errors, both WASAPIOutput and WASAPIInput simple disconnect and stop running. I think this might actually be a separate issue from the original issue here. Maybe I need to move this to a new issue? In any case, it seems like a simple fix. Any fails within the ````while(bRunning && !FAILED(hr))```` loop need to check the error code. If the error code is AUDCLNT_E_DEVICE_INVALIDATED, then instead of completely terminating the thread, it should instead reloop the entire cleanup and then setup of the device again.

Do you guys accept third-party pull requests for this project? I could make the changes and submit a PR if you want

**Comment by Krzmbrzl** [comment]
> Do you guys accept third-party pull requests for this project? I could make the changes and submit a PR if you want

Absolutely! If you could create a PR that fixes this, that'd be very welcome
